ledger/eval/eval.go

```diff
@@ -788,13 +788,89 @@ func StartEvaluator(l LedgerForEvaluator, hdr bookkeeping.BlockHeader, evalOpts
 		return nil, fmt.Errorf("overflowed subtracting rewards for block %v", hdr.Round)
 	}
 
+	if eval.eligibleForIncentives(prevHeader.Proposer) {
+		incentive, _ := basics.NewPercent(proto.MiningPercent).DivvyAlgos(prevHeader.FeesCollected)
+		total, o := basics.OAddA(incentive, prevHeader.Bonus)
+		if o {
+			return nil, fmt.Errorf("overflowed adding bonus incentive %d %d", incentive, prevHeader.Bonus)
+		}
+
+		sink, err := eval.state.lookup(prevHeader.FeeSink)
+		if err != nil {
+			return nil, err
+		}
+		available := sink.AvailableBalance(&proto)
+		total = basics.MinA(total, available)
+
+		err = eval.state.Move(prevHeader.FeeSink, prevHeader.Proposer, total, nil, nil)
+		if err != nil {
+			// Should be impossible, we used AvailableBalance
+			return nil, fmt.Errorf("unable to pay block incentive: %v", err)
+		}
+	}
+
+	// Note their proposal if MiningEnabled.
+	if proto.EnableMining {
+		prp, err := eval.state.Get(prevHeader.Proposer, false)
+		if err != nil {
+			return nil, err
+		}
+		prp.LastProposed = hdr.Round - 1
+		// An account could propose, even while suspended, because of the 320 round lookback.
+		if prp.Suspended() {
+			prp.Status = basics.Online
```

Comment:
>  it should un-suspend the proposer in 320 rounds

That's what actually happens.  When an account's status is changed in, say, round 2000, it doesn't get to start voting until 2320. Similarly, if an account is suspended in round 3000, then proposes in 3100, there will be a 100 round "bubble" in which is can't vote.  It won't be able to vote from 3320 until 3420.

For comparison, consider that a `keyreg` transaction performs the `Status` update the same way, by setting the account's status "immediately".  The machinery to not use that status until 320 rounds later happens in the online accounts tracker, and because `agreement` is always really asking "what was the online stake 320 rounds ago, and who was online"?

Reply:
Ah ok, I missed that. I'll have to look closer to convince myself that's how it works. Resolving